### PR TITLE
more work on projector, countdowns, clos

### PR DESCRIPTION
**Pull request by FinnStutzenstein**
What is still TODO:

- Comments
- ~~The current list of spekaers is buggy. Do not use it, it might break the projector elements~~
- ~~Clock- and~~ countdownslide: Sync with servertime.
- CLOS reference (currently discussing with norman/oscar via email). Another reason not to use it.
- There might be some bugs..

Someone can take this code and make some nice slides. Someone may merge this, it won't break anything not-projector-related. But if this is merged, other people can work on slides

@emanuelschuetze @tsiegleauq decide. I'll work one the above meantioned things next week.

### CODE REVIEW

**Comment by FinnStutzenstein** [comment]
There's more:

- The docs (especially everything `ProjectorService` related) are completly outdated, becaused I changed the structure there.
- Messages needs to have a list-view (this can be done by someone. This is not directly related to the projector stuff. It's the same as categories, statute paragraphs, and so on)
- Countdowns needs to be manageable
- I need a service, that can be injected to a slides and provide the current projector id, where the slide is projected on.

**Comment by FinnStutzenstein** [comment]
Info: There is an update. CLOS should work (except for the actual content of the slide/overlay)

File: openslides/core/serializers.py [hunk 1]
```diff
@@ -78,7 +78,7 @@ class ProjectorSerializer(ModelSerializer):
     """
 
     elements = JSONSerializerField(validators=[elements_validator])
-    elements_preview = JSONSerializerField(validators=[elements_array_validator])
+    elements_preview = JSONSerializerField(validators=[elements_validator])
```
Comment (ostcar) [inline openslides/core/serializers.py]: This seams wrong. The preview field is an array. Why did you change this?

File: openslides/core/views.py [hunk 2]
```diff
@@ -164,20 +164,28 @@ def project(self, request, pk):
...
+
+        If `delete_last_history_element` is True, the last element is deleted.
+        Note: You cannot give `append_to_history` and `delete_last_history_element`
+        at the same time.
         """
         projector = self.get_object()
         elements = request.data.get("elements")
         preview = request.data.get("preview")
         history_element = request.data.get("append_to_history")
+        delete_last_history_element = request.data.get("delete_last_history_element", False)
```
Comment (ostcar) [inline openslides/core/views.py]: Are you sure this has to be part of this view or should it be a separate view?

File: openslides/core/serializers.py (hunk 1 above)
Comment (FinnStutzenstein) [reply openslides/core/serializers.py]: Its an array of single elements. The preview just supports one element at a time, so you have a list of elements, that will be projected.

Comment (FinnStutzenstein) [reply openslides/core/serializers.py]: You can have a look in the client, then it might be clear (note, that the "append to preview" button is missing, but you can use next/previous)

**Comment by FinnStutzenstein** [comment]
More todos:

- add to preview button in projection dialog
- mobile friendy projection

**Comment by emanuelschuetze** [comment]
@FinnStutzenstein I added some template and CSS work for a first projector container layout with motion slide. Question: How can I iterate over submitters in motion slide template?

File: client/src/app/slides/motions/motion/motions-motion-slide.component.html [hunk 3]
```diff
@@ -1,4 +1,38 @@
...
+        <div *ngIf="data.data.recommendation">
+            <h3 translate>Recommendation</h3> <!-- TODO: use config 'motions_recommendations_by'-->
+            {{ data.data.recommendation }}
+        </div>
+
+        <!-- Submitters -->
+        <h3 translate>Submitters</h3>
+        <div *ngFor="let submitter of data.data.submitter">
+            <!-- TODO: order by weight-->
+            {{ submitter }}
```
Comment (emanuelschuetze) [inline client/src/app/slides/motions/motion/motions-motion-slide.component.html]: @FinnStutzenstein Any idea how I can get the submitter name?

File: client/src/app/slides/motions/motion/motions-motion-slide.component.html [hunk 4]
```diff
@@ -1,4 +1,38 @@
...
+    <div id="sidebox" *ngIf="data.data.show_meta_box">
+        <!-- State -->
+        <h3 translate>State</h3>
+        {{ data.data.state }}
+        <span *ngIf="data.data.state_extension">{{ data.data.state_extension }}</span>
+
+        <!-- Recommendation -->
+        <div *ngIf="data.data.recommendation">
+            <h3 translate>Recommendation</h3> <!-- TODO: use config 'motions_recommendations_by'-->
+            {{ data.data.recommendation }}
```
Comment (emanuelschuetze) [inline client/src/app/slides/motions/motion/motions-motion-slide.component.html]: motions with recommendation failed with server error #4190

File: client/src/app/slides/motions/motion/motions-motion-slide-data.ts [hunk 5]
```diff
@@ -1,3 +1,16 @@
...
+    amendment_paragraphs: object[];
+    is_child: boolean;
+    show_meta_box: boolean;
+    reason: string;
+    modified_final_version: string;
+    state: string;
+    state_extension: string;
+    recommendation: string;
+    recommendation_extension: string;
+    submitter: object[];
```
Comment (emanuelschuetze) [inline client/src/app/slides/motions/motion/motions-motion-slide-data.ts]: @FinnStutzenstein is this type `object[]` correct?

File: client/src/app/slides/motions/motion/motions-motion-slide.component.html (hunk 3 above)
Comment (FinnStutzenstein) [reply client/src/app/slides/motions/motion/motions-motion-slide.component.html]: The server gives us a "small user representation". Use all given fields to make full/short name. If fields are missing, add them in the server.

The user model does have implementations for full/short name. Move them out of the model, so you can use them from the model and the slides. You must just add a few parameters for all needed things (username, firstname, title, ...)

File: client/src/app/slides/motions/motion/motions-motion-slide-data.ts (hunk 5 above)
Comment (FinnStutzenstein) [reply client/src/app/slides/motions/motion/motions-motion-slide-data.ts]: Nope. Please completly specify the type
I think the type is currently
{title:string;
firstname:string;
lastname:string;}[]

Also more things: not everything is always send by the server: the reason might not be present, if the config "show reason" is disabled. Mark all fieds with a '?', that might be missing. Check the servercode. All fields that have an 'if' around it are optional. In the slide you always need to check, if some data is present before using it.

File: client/src/app/slides/motions/motion/motions-motion-slide.component.html (hunk 4 above)
Comment (FinnStutzenstein) [reply client/src/app/slides/motions/motion/motions-motion-slide.component.html]: This might be possible, I didn't check the code there. If you can repair it, go on! We (as the client devs) know, what data we need, so we must puzzle it together from the given data on the server.

Comment (emanuelschuetze) [reply client/src/app/slides/motions/motion/motions-motion-slide.component.html]: Fixed

File: client/src/app/slides/motions/motion/motions-motion-slide.component.html (hunk 3 above)
Comment (emanuelschuetze) [reply client/src/app/slides/motions/motion/motions-motion-slide.component.html]: Will be fixed in #4194

**Comment by emanuelschuetze** [comment]
Rebased on top of master. Motion and user slide finished.

**Comment by emanuelschuetze** [comment]
@tsiegleauq Please review

**Comment by emanuelschuetze** [comment]
Travis fixed.

File: client/src/app/core/services/projector.service.ts [hunk 6]
```diff
@@ -44,10 +62,28 @@ export class ProjectorService extends OpenSlidesComponent {
      * @param obj The object in question
      * @return All projectors, where this Object is projected on
      */
-    public getProjectorsWhichAreProjecting(obj: Projectable): Projector[] {
-        return this.DS.getAll<Projector>('core/projector').filter(projector => {
-            return projector.isElementShown(obj.getNameForSlide(), obj.getIdForSlide());
-        });
+    public getProjectorsWhichAreProjecting(obj: Projectable | ProjectorElementBuildDeskriptor): Projector[] {
```
Comment (tsiegleauq) [inline client/src/app/core/services/projector.service.ts]: Better naming could be
`getActiveProjectors`

File: client/src/app/core/services/projector.service.ts [hunk 7]
```diff
@@ -103,4 +205,60 @@ export class ProjectorService extends OpenSlidesComponent {
...
+            throw new Error('THis projectorelement cannot be mapped to a model');
+        }
+        const identifiers = element.getIdentifiers();
+        if (!identifiers.includes('name') || !identifiers.includes('name')) {
+            throw new Error('To map this element to a model, a name and id is needed.');
+        }
+        return this.DS.get<T>(element.name, element.id);
+    }
+
+    public async projectNextSlide(projector: Projector): Promise<void> {
```
Comment (tsiegleauq) [inline client/src/app/core/services/projector.service.ts]: many methods have the prefix "project".
Would it be misleading to just name the methos `nextSlide` and `previewSlide`, `previousSlide`

File: client/src/app/site/base/projectable.ts [hunk 8]
```diff
@@ -1,32 +1,34 @@
...
-    getIdForSlide(): number | null;
+    getTitle(): string;
+}
 
-    /**
-     * The slide's name
-     */
-    getNameForSlide(): string;
+export function isProjectable(obj: any): obj is Projectable {
+    return (<Projectable>obj).getSlide !== undefined;
```
Comment (tsiegleauq) [inline client/src/app/site/base/projectable.ts]: should maybe be `(<Projectable>obj).getSlide()`
`getSlide` would return the body of a function, that will never be undefined

**Comment by tsiegleauq** [comment]
@emanuelschuetze if merging is (still) urgent, please rebase and merge.

**Comment by emanuelschuetze** [comment]
Rebased on top of master. Resolved conflict. Ready to merge. 
@FinnStutzenstein Please read the remarks and fix them in a new PR. Thanks.

File: openslides/users/projector.py [hunk 9]
```diff
@@ -12,8 +16,22 @@
...
     User slide.
+
+    The returned dict can contain the following fields:
+    * user
     """
-    return {"error": "TODO"}
+    user_id = element.get("id")
+
+    if user_id is None:
+        return {"error": "id is required for user slide"}
```
Comment (ostcar) [inline openslides/users/projector.py]: Please use the ProjectorElementException inserted with #4194 

See also line 31 in this file

File: openslides/users/projector.py [hunk 10]
```diff
@@ -12,8 +16,22 @@
...
+
+    try:
+        user = all_data["users/user"][user_id]
+    except KeyError:
+        raise ProjectorElementException(f"user with id {user_id} does not exist")
+
+    return_value = {"user": get_user_name(all_data, user["id"])}
+    return return_value
```
Comment (ostcar) [inline openslides/users/projector.py]: The variable return_value is not needed. Please return the value directly

**Comment by FinnStutzenstein** [comment]
I'll fix this thuesday when I do my marked todos
